update to scala 3

rajcspsg opened the pull request:
Hi,

I'm trying to update the plugin to provide support for scala 3.0.0-RC1 version.
I've created a sample project https://github.com/rajcspsg/scala_3_mvn_demo/blob/master/pom.xml#L1-L56. But when I do `mvn clean compile`  I get CNFE.
The complete error trace is here https://gist.github.com/rajcspsg/7ca4cc036b35a867f06fab957042b214.

I'm running out of ideas and  need help from this community how to fix this.

davidB reviewed (commented):
The reformat and (unrelated) changes in the `pom.xml` add lot noise in the PR.

davidB commented:
About the CNFE, I'm not an zinc/sbt expert, but I don't know if zinc can run with scala3. I mean 
1. Can zinc be executed with scala3 in its own classpath (in place of scala2) ?
2. Can zinc compile scala3 code ?

davidB commented on src/main/java/scala_maven/ScalaMojoSupport.java:
```diff
@@ -783,47 +790,81 @@ private String getToolClasspath() throws Exception {
     return options;
   }
 
+  private String getScala3ArtifactId(String a) {
+    String _scalaVersionNStr = _scalaVersionN.toString();
+    // int indexOf_ = _scalaVersionNStr.indexOf("-");
+    return _scalaVersionN.major == 3 ? a + "_" + _scalaVersionNStr : a;
+  }
+
```
I don't understand the purpose of this function and it's usage like
```
        _scalaVersionN.major == 3
            ? getScala3ArtifactId(SCALA3_LIBRARY_ARTIFACTID)
            : SCALA_LIBRARY_ARTIFACTID;
  ```

davidB replied:
I looked at scala3 on search.maven.org, I understand now the change, the naming convention has changed again (worst IMHO). So to support scala3 naming convention a refactor of the code should be done. The scala-maven-plugin should work in `incremental` mode (zinc) and `all` mode (scala compiler), both mode are impacted by thoses changes in naming.

davidB commented on src/main/java/scala_maven/ScalaMojoSupport.java:
```diff
@@ -783,47 +790,81 @@ private String getToolClasspath() throws Exception {
     }
-    return getArtifactJar(
-        getScalaOrganization(), SCALA_LIBRARY_ARTIFACTID, findScalaVersion().toString());
+    return getArtifactJar(getScalaOrganization(), scalaLibrary, findScalaVersion().toString());
```
I guess the value of `scalaLibrary` is not right one for scala 3, because `getScala3ArtifactId` will return an artifact name with a version number => not found

davidB commented:
Thanks for contributing to scala3 support.

SethTisue commented:
perhaps @eed3si9n can help here?

sjrd commented:
@adpi2 may be able to answer the questions about Zinc.

adpi2 commented:
> Can zinc be executed with scala3 in its own classpath (in place of scala2) ?

Possibly yes. 
With the intercompat between Scala 2.13 and Scala 3, it should be possible to run `zinc_2.13` with `scala3-compiler_3` and `scala3-library_3` in the same class path. That is not how we do it in sbt but it should work.

> Can Zinc compile scala3 code ?

Yes with some precautions.

The ScalaInstance class loader must contain the `scala3-compiler` jar and all its dependencies. In particular it now depends directly on `compiler-interface`, [the Zinc interface](https://mvnrepository.com/artifact/org.scala-sbt/compiler-interface), and the class loader for that jar must be shared between the `ScalaMavenPlugin` and the `ScalaInstance`.

So the code [here](https://github.com/davidB/scala-maven-plugin/blob/d7d19a2e98e7661d30fadb726fb0e91177eee5b6/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L96-L103) would probably looks like this:
```java
new ScalaInstance(
    scalaVersion.toString(), // version
    new URLClassLoader(
        allScala3CompilerJars,
        SbtIncrementalCompiler.class.getClassLoader()
    ),
    ...
)
```

Also the Scala 3 compiler bridge that is fetched [here](https://github.com/davidB/scala-maven-plugin/blob/d7d19a2e98e7661d30fadb726fb0e91177eee5b6/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L111) is now called [scala3-sbt-bridge](https://mvnrepository.com/artifact/org.scala-lang/scala3-sbt-bridge/3.0.0-RC1) and a new version is released for each Scala 3 version (rather than for each Zinc version)

rajcspsg commented:
Thanks @adpi2 for your comments. I will try your suggestion today and test and scala 3.

rajcspsg commented on pom.xml:
```diff
@@ -339,6 +341,10 @@
         <plugin>
           <artifactId>maven-compiler-plugin</artifactId>
           <version>3.8.1</version>
+          <configuration>
+            <source>1.8</source>
+            <target>1.8</target>
+          </configuration>
```
@davidB without this config, the tests are failing in my local

rajcspsg commented on src/main/java/scala_maven/ScalaMojoSupport.java:
```diff
@@ -283,9 +283,12 @@ public String getScalaOrganization() {
    * @return a {@link Artifact} for the Scala Compiler.
    */
   final Artifact scalaCompilerArtifact(String scalaVersion) {
+    System.out.println("scalaCompilerArtifact scalaVersion=" + scalaVersion);
```
i'mm remove this

rajcspsg commented:
@adpi2 I tried your suggestion. the compilation is successful for scala 2 code. For example below code compiles fine.

```scala
object DemoApp extends App {
  println("Hello World!!!")
}

```

But when I tried to compile scala 3 code I get below error - 

>[INFO] compiling 2 Scala sources to /Users/rajkumar.natarajan/Documents/Coding/OS/scala_3_mvn_demo/target/classes ...
>[ERROR] /Users/rajkumar.natarajan/Documents/Coding/OS/scala_3_mvn_demo/src/main/scala/Test.scala:1: ';' expected but ':' found.
>[ERROR] one error found

The sample scala 3 code is -

```scala
object happyBirthday extends App:
    println("Happy Birthday")
```

I even tried adding all the compile dependencies of [scala3-compiler](https://mvnrepository.com/artifact/org.scala-lang/scala3-compiler_3.0.0-RC2/3.0.0-RC2) (it is not in this commit) but still I get the same error -

```new URL[] {
                  libraryJar.toURI().toURL(),
                  interfacesJar.toURI().toURL(),
                  compilerJar.toURI().toURL(),
                  new File(
                          "https://repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar")
                      .toURI()
                      .toURL(),
                  new File(
                          "https://repo1.maven.org/maven2/org/scala-lang/tasty-core_3.0.0-RC2/3.0.0-RC2/tasty-core_3.0.0-RC2-3.0.0-RC2.jar")
                      .toURI()
                      .toURL(),
                  new File(
                          "https://repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/9.1.0-scala-1/scala-asm-9.1.0-scala-1.jar")
                      .toURI()
                      .toURL()
                }
```


Any idea what might be the problem?

adpi2 commented on src/main/java/sbt_inc/SbtIncrementalCompiler.java:
```diff
@@ -304,12 +316,10 @@ private File getCompiledBridgeJar(ScalaInstance scalaInstance, Log mavenLogger)
       // compile and install
       RawCompiler rawCompiler = new RawCompiler(scalaInstance, ClasspathOptionsUtil.auto(), logger);
 
-      File bridgeSources =
-          resolver.getJar(SBT_GROUP_ID, bridgeArtifactId, zincVersion, "sources").getFile();
+      File bridgeSources = resolver.getJar(groupId, bridgeArtifactId, version, "sources").getFile();
 
```
I forgot to say that the Scala 3 compiler bridge (`scala3-sbt-bridge`) is pure java. That means you don't need to fetch the sources and to compile them. You can use the jar files directly.

So probably the code will become:

```
if (isScala3) {
  return resolver.getJar(groupId, bridgeArtifactId, version).getFile(); // fetch the jar and use as it is
} else {
  // fetch the sources
  File bridgeSources = resolver.getJar(groupId, bridgeArtifactId, version, "sources").getFile();

  // compile the sources
  // build a jar file
}

```

slandelle replied:
> I forgot to say that the Scala 3 compiler bridge (scala3-sbt-bridge) is pure java. That means you don't need to fetch the sources and to compile them. You can use the jar files directly.

![4318779](https://user-images.githubusercontent.com/849197/113113269-682c6400-920a-11eb-9bc1-e01a77516dec.jpg)

slandelle commented:
I'm super swamped atm, sorry :(
Thanks a lot to all of you for helping with this!

adpi2 commented on src/main/java/sbt_inc/SbtIncrementalCompiler.java:
```diff
@@ -111,10 +112,14 @@ public SbtIncrementalCompiler(
 
+    URLClassLoader classLoader =
+        scalaVersion.major == 3
+            ? new URLClassLoader(urls, SbtIncrementalCompiler.class.getClassLoader())
```
I am not 100% sure but I think the `SbtIncrementalCompiler.class.getClassLoader()` contains the Scala 2 compiler jars. Those will hide the Scala 3 compiler classes and prevent us from compiling in Scala 3.

What solution do we have?

## 1. Use a Layered class loader

Invoke `SbtIncrementalCompiler` in a layered class loader.

![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBcbiAgICBBW3NjYWxhLW1hdmVuLXBsdWdpbi5qYXI8YnIvPnNjYWxhLWNvbXBpbGVyLmphcjxici8-c2NhbGEtbGlicmFyeS5qYXI8YnIvPi4uLl0gLS0-fHBhcmVudHwgQlxuICAgIEJbY29tcGlsZXItaW50ZXJmYWNlLmphcl0gLS0-fHBhcmVudHxDXG4gICAgQ1tudWxsXVxuICAgIERbc2NhbGEzLWNvbXBpbGVyLmphcjxici8-c2NhbGEzLWxpYnJhcnkuamFyPGJyLz4uLi5dIC0tPiB8cGFyZW50fEIiLCJtZXJtYWlkIjp7fSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

To build the Scala 3 compiler class loader you can get the `compiler-interface.jar` class loader:

```scala
? new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader(xsbti.Reporter.class.getClassLoader()))
```

## 2. Create a filtered class loader

Encapsulate `SbtIncrementalCompiler.class.getClassLoader()` in a custom `Scala3CompilerLoader` that resolves the `xsbti.*` classes (classes from `compiler-interface.jar`)  from the `SbtIncrementalCompiler.class.getClassLoader()`.

```java
? new Scala3CompilerLoader(urls, xsbti..class.getClassLoader(xsbti.Reporter.class.getClassLoader()))
```
Where:

```java
class Scala3CompilerLoader extends URLClassLoader {
  private final ClassLoader sbtLoader;

  public CompilerClassLoader(URL[] urls, ClassLoader sbtLoader) {
    super(urls, null);
    this.sbtLoader = sbtLoader;
  }

  @Override
  public Class<?> loadClass(String className, boolean resolve) throws ClassNotFoundException {
    if (className.startsWith("xsbti.")) {
      // We can't use the loadClass overload with two arguments because it's
      // protected, but we can do the same by hand (the classloader instance
      // from which we call resolveClass does not matter).
      Class<?> c = sbtLoader.loadClass(className);
      if (resolve)
        resolveClass(c);
      return c;
    } else {
      return super.loadClass(className, resolve);
    }
  }
}
```

rajcspsg replied:
@adpi2 `new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader(xsbti.Reporter.class.getClassLoader()))` is not compiling. you meant `new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader())`?

rajcspsg replied:
@adpi2 `new Scala3CompilerLoader(urls, xsbti..class.getClassLoader(xsbti.Reporter.class.getClassLoader()))` is also gives compile error. could you please take a look at my latest commit?

adpi2 replied:
Your latest commit looks good

adpi2 commented:
@rajcspsg Probably your test is not working because the Scala 2 compiler in `SbtIncrementalCompiler.class.getClassLoader()` is invoked instead of the Scala 3 compiler.

davidB commented:
Happy to see progress / discussion on this PR.
Another expected point to add to the PR, a test under `src/it/test_scala3` (with your current sample app). To invoke the only this test:
```
./mvnw integration-test -Dinvoker.test=test_scala3
```

When setup it should also speedup your current workflow that should look like  `mvn install && pushd ../sample_scala3 && mvn compile && popd`

rajcspsg commented:
seems like I'm getting closer. I still get this [error](https://gist.github.com/rajcspsg/b46c921d30168bab1563a16cbfeb963d). probably I will check this later tomorrow

rajcspsg commented:
@adpi2 I need your help and I'm running out of ideas to debug here.

Below is the error I'm getting when I compile the code.

>[INFO] compiling 1 Scala source to /Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/classes ...
>[ERROR] ## Exception when compiling 1 sources to /Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/classes
>java.lang.NoClassDefFoundError: scala/MatchError
>dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:21)
>sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
>sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
>scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
>sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
>sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
>sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
>sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
>sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
>sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
>sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
>sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
>sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
>sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
>sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
>sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
>sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
>sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
>sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
>sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
>sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
>sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
>sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
>sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
>sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
>sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
>sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
>sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
>sbt_inc.SbtIncrementalCompiler.compile(SbtIncrementalCompiler.java:237)
>scala_maven.ScalaCompilerSupport.incrementalCompile(ScalaCompilerSupport.java:328)
>scala_maven.ScalaCompilerSupport.compile(ScalaCompilerSupport.java:121)
>scala_maven.ScalaCompilerSupport.doExecute(ScalaCompilerSupport.java:96)
>scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:473)
>org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
>org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
>org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
>org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
>org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
>org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
>org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
>org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
>org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
>org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
>org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
>org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
>org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
>org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
>sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>java.lang.reflect.Method.invoke(Method.java:498)
>org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
>org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
>org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
>org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)

When I checked the urls list, again it used 2.13 version of scala jar files for compilation.

Could you please help me what might be wrong here?

adpi2 commented on src/main/java/sbt_inc/SbtIncrementalCompiler.java:
```diff
@@ -89,18 +91,35 @@ public SbtIncrementalCompiler(
-                new URL[] {
-                  libraryJar.toURI().toURL(),
-                  reflectJar.toURI().toURL(),
-                  compilerJar.toURI().toURL()
-                }), // loader
+            classLoader, // loader
```
The error indicates that `scala-library` is missing from the classpath.

The `ScalaInstance` constructor that is used here is deprecated. You should use the one defined here: https://github.com/sbt/zinc/blob/a096e0d32dd46c0be4c831a37b497947e7479b3a/internal/zinc-classpath/src/main/scala/sbt/internal/inc/ScalaInstance.scala#L34-L43

Make sure that the different lists of jars are correct:
`compilerJars` should contain:
- `scala3-compiler`
- `scala3-interfaces`
- ` tasty-core`
- ...

`libraryJars` should contain:
- `scala3-library`
-  `scala-library` (the scala3-library depends on the Scala 2.13 library)

`allJars` should contain all of them.

Note that each list of jars is associated with a classloader:
```
val loaderLibraryOnly = new Scala3CompilerLoader(libraryJars, xsbti.Reporter.class.getClassLoader())
val loaderCompilerOnly = new URLClassLoader(compilerJars, loaderLibraryOnly)
val loader = loaderCompilerOnly
```

adpi2 replied:
As you can see the logic to configure the Scala 3 compiler is quite different from Scala 2. Maybe you should have 2 different constructors.

davidB commented:
@rajcspsg, I merged into `master` https://github.com/davidB/scala-maven-plugin/pull/451 the update to zinc 1.5.0, that seems to include feature for scala3 support by @adpi2 .
Try to rebase your fork, branch, PR.

rajcspsg commented:
@davidB @adpi2 I've incorporated your latest comments and used zinc version 1.5.0. 
After this I see test_incremental_compile is failing (It is also failing in the current master branch as well. I think it is due to zinc 1.5.0).
test_scala3 tests is also failing. But when I tested the other project https://github.com/rajcspsg/scala_3_mvn_demo using `4.4.2-SNAPSHOT` version then `mvn clean install` command is success. https://imgur.com/a/8DlME8m.

One more thing I need to address is `mvn scala:console` is failing with error `CNFE: scala.tools.nsc.MainGenericRunner`

I tried to use `I tried to use dotty.tools.dotc.Run` but got another error 
>wrap: java.lang.NoSuchMethodException: dotty.tools.dotc.Run.main([Ljava.lang.String;)

Any idea?

rajcspsg commented:
I figured out why test for test_scala3 is failing.

The command executed by .JavaMainCallerByFork class is-

```/Users/rajkumar.natarajan/.sdkman/candidates/java/8.0.232-zulu/bin/java -Xbootclasspath/a:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-library_3.0.0-RC2/3.0.0-RC2/scala3-library_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-compiler_3.0.0-RC2/3.0.0-RC2/scala3-compiler_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-interfaces/3.0.0-RC2/scala3-interfaces-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/modules/scala-asm/9.1.0-scala-1/scala-asm-9.1.0-scala-1.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/tasty-core_3.0.0-RC2/3.0.0-RC2/tasty-core_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar:/Users/rajkumar.natarajan/.m2/repository/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar -DwordWithSpace=hello world -classpath /Users/rajkumar.natarajan/.m2/repository/net/alchim31/maven/scala-maven-plugin/4.4.2-SNAPSHOT/scala-maven-plugin-4.4.2-SNAPSHOT.jar scala_maven_executions.MainWithArgsInFile dotty.tools.dotc.Main /private/var/folders/9w/cpm7qr_d64sdxv15r99cq8h17jkthy/T/scala-maven-1158848394192223510.args
```

The contents of arg file /private/var/folders/9w/cpm7qr_d64sdxv15r99cq8h17jkthy/T/scala-maven-1158848394192223510.args is -

```
$ cat /private/var/folders/9w/cpm7qr_d64sdxv15r99cq8h17jkthy/T/scala-maven-1158848394192223510.args
-dependencyfile
/Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/.main_scala_dependencies
-classpath
/Users/rajkumar.natarajan/.m2/repository/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-library_3.0.0-RC2/3.0.0-RC2/scala3-library_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/modules/scala-asm/9.1.0-scala-1/scala-asm-9.1.0-scala-1.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-interfaces/3.0.0-RC2/scala3-interfaces-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/tasty-core_3.0.0-RC2/3.0.0-RC2/tasty-core_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar:/Users/rajkumar.natarajan/.m2/repository/org/scala-lang/scala3-compiler_3.0.0-RC2/3.0.0-RC2/scala3-compiler_3.0.0-RC2-3.0.0-RC2.jar:/Users/rajkumar.natarajan/.m2/repository/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar
-d
/Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/classes
/Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/src/main/scala/TestScalaApp.scala
```

The dependency file `/Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/.main_scala_dependencies` is not present. I don't know why it is not.
When I removed those lines manually `-dependencyfile` and `/Users/rajkumar.natarajan/Documents/Coding/OS/scala-maven-plugin/target/it/test_scala3/target/.main_scala_dependencies`  and run the same command then it is successful.

@davidB Any idea why that file is not present in that expected directory?

rajcspsg commented:
Only 2 outstanding issues are pending -

1. I've incorporated your latest comments and used zinc version 1.5.0.
After this I see test_incremental_compile is failing (It is also failing in the current master branch as well. I think it is due to zinc 1.5.0).

2. `mvn scala:console` is failing with error `CNFE: scala.tools.nsc.MainGenericRunner`

adpi2 commented on src/main/java/scala_maven/ScalaCompilerSupport.java:
```diff
@@ -276,6 +289,104 @@ void setLastSuccessfullTS(long v) throws Exception {
+        new File[] {
+          getCompilerJar(scalaVersion),
+          getArtifactJar(scalaOrg, SCALA3_INTERFACES_ARTIFACTID, version),
+          getArtifactJar(scalaOrg, "tasty-core_" + version, version),
```
This will not work for Scala 3.0.0 onward because the binary version will be "3".
```suggestion
          getArtifactJar(scalaOrg, "tasty-core_" + binaryVersion, version),
```

`binaryVersion` should be `"3.0.0-x"` if Scala version is 3.0.0-x or 3.0.0-x-bin-y
but it should be `"3"` if Scala version >= 3.0.0  (for instance 3.0.0, 3.0.1-RC1, 3.0.1, 3.1.0...)

adpi2 commented on src/main/java/scala_maven/ScalaMojoSupport.java:
```diff
@@ -783,47 +789,89 @@ private String getToolClasspath() throws Exception {
     return options;
   }
 
+  private String getScala3ArtifactId(String a) {
+    String _scalaVersionNStr = _scalaVersionN.toString();
+    // int indexOf_ = _scalaVersionNStr.indexOf("-");
+    return _scalaVersionN.major == 3 ? a + "_" + _scalaVersionNStr : a;
```
Same here _scalaVersionNStr should be replaced by the binary version which will be "3" starting from Scala 3.0.0

adpi2 replied:
This will not work for Scala 3.0.0 onward because the binary version will be "3".
```scala
         val binaryVersion =
            if(scalaVersionN.major == 3 && scalaVersionN.minor == 0 && scalaVersionN.bugfix == 0 && scalaVersionN.modifier != null) scalaVersionN.toString
           else "3"
```

`binaryVersion` should be `"3.0.0-x"` if Scala version is 3.0.0-x or 3.0.0-x-bin-y
but it should be `"3"` if Scala version >= 3.0.0  (for instance 3.0.0, 3.0.1-RC1, 3.0.1, 3.1.0...)

adpi2 commented:
@rajcspsg Thanks a lot for your work on this!

> 1. I've incorporated your latest comments and used zinc version 1.5.0.
After this I see test_incremental_compile is failing (It is also failing in the current master branch as well. I think it is due to zinc 1.5.0).

Weirdly, things were working well on my computer, so I spent quite some time figuring out what is going on in the CI. Here are my findings:
- this branch fails to compile the `sbt-compiler-bridge_2.12` and it creates a stale `~/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.5.0-bin_2.12.12__52.0-1.5.0_20210404T034433.jar`
- if you remove that file you should be able to run the `master` branch successfully
- to reproduce the error in the `update_for_scala3` branch you will need to remove the jar file again. You should see the following logs:
```text
[INFO] [INFO] Compiler bridge file: /home/piquerez/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.5.0-bin_2.12.12__52.0-1.5.0_20210404T034433.jar
[INFO] [INFO] Compiler bridge file is not installed yet
[INFO] Usage: scalac <options> <source files>
[INFO] where possible standard options include:
[INFO]   -Dproperty=value                     Pass -Dproperty=value directly to the runtime system.
[INFO]   -J<flag>                             Pass <flag> directly to the runtime system.
[INFO]   -P:<plugin>:<opt>                    Pass an option to a plugin
[INFO]   -X                                   Print a synopsis of advanced options.
[INFO]   -bootclasspath <path>                Override location of bootstrap class files.
[INFO]   -classpath <path>                    Specify where to find user class files.
[INFO]   -d <directory|jar>                   destination for generated classfiles.
[INFO]   -dependencyfile <file>               Set dependency tracking file.
[INFO]   -deprecation                         Emit warning and location for usages of deprecated APIs.
[INFO]   -encoding <encoding>                 Specify character encoding used by source files.
[INFO]   -explaintypes                        Explain type errors in more detail.
[INFO]   -extdirs <path>                      Override location of installed extensions.
[INFO]   -feature                             Emit warning and location for usages of features that should be imported explicitly.
[INFO]   -g:<level>                           Set level of generated debugging info. Choices: (none,source,line,vars,notailcalls), default: vars.
[INFO]   -help                                Print a synopsis of standard options
[INFO]   -javabootclasspath <path>            Override java boot classpath.
[INFO]   -javaextdirs <path>                  Override java extdirs classpath.
[INFO]   -language:<_,feature,-feature>       Enable or disable language features: `_' for all, `-language:help' to list choices.
[INFO]   -no-specialization                   Ignore @specialize annotations.
[INFO]   -nobootcp                            Do not use the boot classpath for the scala jars.
[INFO]   -nowarn                              Generate no warnings.
[INFO]   -opt:<_,optimization,-optimization>  Enable optimizations: `_' for all, `-opt:help' to list choices.
[INFO]   -opt-inline-from:<patterns>          Patterns for classfile names from which to allow inlining, `help` for details.
[INFO]   -opt-warnings:<_,warning,-warning>   Enable optimizer warnings: `_' for all, `-opt-warnings:help' to list choices.
[INFO]   -print                               Print program with Scala-specific features removed.
[INFO]   -release <<release>>                 Compile for a specific version of the Java platform. Supported targets: 6, 7, 8, 9
[INFO]   -sourcepath <path>                   Specify location(s) of source files.
[INFO]   -target:<target>                     Target platform for object files. All JVM 1.5 - 1.7 targets are deprecated. Choices: (jvm-1.5,jvm-1.6,jvm-1.7,jvm-1.8), default: jvm-1.8.
[INFO]   -toolcp <path>                       Add to the runner classpath.
[INFO]   -unchecked                           Enable additional warnings where generated code depends on assumptions.
[INFO]   -uniqid                              Uniquely tag all identifiers in debugging output.
[INFO]   -usejavacp                           Utilize the java.class.path in classpath resolution.
[INFO]   -usemanifestcp                       Utilize the manifest in classpath resolution.
[INFO]   -verbose                             Output messages about what the compiler is doing.
[INFO]   -version                             Print product version and exit.
[INFO]   @<file>                              A text file containing compiler arguments (options and source files)
[INFO] 
[INFO] Deprecated settings:
[INFO]   -optimise                            Compiler flag for the optimizer in Scala 2.11
[INFO]                                          deprecated: In 2.12, -optimise enables -opt:l:inline -opt-inline-from:**. Check -opt:help for using the Scala 2.12 optimizer.
[INFO] 
[INFO] [INFO] Compiler bridge installed
```
It indicates that the compiler is not invoked correctly here:
https://github.com/rajcspsg/scala-maven-plugin/blob/d42d0b5c923a55e68ff6e9d33891c1651606cb1c/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L274

> 2. mvn scala:console is failing with error CNFE: scala.tools.nsc.MainGenericRunner

In a Scala 2.12 project? If yes, this is probably the same problem than  `1.`.

adpi2 commented on src/main/java/scala_maven/ScalaCompilerSupport.java:
```diff
@@ -276,6 +289,104 @@ void setLastSuccessfullTS(long v) throws Exception {
+
+    File[] compilerJars =
+        new File[] {
+          getCompilerJar(scalaVersion),
+          getArtifactJar(scalaOrg, SCALA3_INTERFACES_ARTIFACTID, version),
+          getArtifactJar(scalaOrg, "tasty-core_" + version, version),
+          getArtifactJar("org.scala-lang.modules", "scala-asm", "9.1.0-scala-1"),
+          getArtifactJar("org.scala-sbt", "compiler-interface", "1.3.5"),
+          getArtifactJar("org.jline", "jline-reader", "3.19.0"),
+          getArtifactJar("org.jline", "jline-terminal", "3.19.0"),
+          getArtifactJar("org.jline", "jline-terminal-jna", "3.19.0")
```
Rather than fetching all jars individually, it is probably simpler to fetch the compiler and all its dependencies in one call:
```scala
resolver.getJarAndDependencies(scalaOrg, scala3compilerId, version, null).map(_.file).toArray
```

adpi2 replied:
then you should be able to filter `scala3-library` and `scala-libary` from compilerJars:

```scala
val libraryJars = compilerJars.filter(jar => jar.getName.contains("scala3-library") || jar.getName.contains("scala-library))
```

rajcspsg replied:
I address this in my last commit.

davidB commented:
Do not forgot to complete the contributor sections in root pom.xml with your names (@rajcspsg & @adpi2).
Thank again for your time and hard work.

rajcspsg commented:
> @rajcspsg Thanks a lot for your work on this!
> 
> > 1. I've incorporated your latest comments and used zinc version 1.5.0.
> >    After this I see test_incremental_compile is failing (It is also failing in the current master branch as well. I think it is due to zinc 1.5.0).
> 
> Weirdly, things were working well on my computer, so I spent quite some time figuring out what is going on in the CI. Here are my findings:
> 
> * this branch fails to compile the `sbt-compiler-bridge_2.12` and it creates a stale `~/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.5.0-bin_2.12.12__52.0-1.5.0_20210404T034433.jar`
> * if you remove that file you should be able to run the `master` branch successfully
> * to reproduce the error in the `update_for_scala3` branch you will need to remove the jar file again. You should see the following logs:
> 
> ```
> [INFO] [INFO] Compiler bridge file: /home/piquerez/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.5.0-bin_2.12.12__52.0-1.5.0_20210404T034433.jar
> [INFO] [INFO] Compiler bridge file is not installed yet
> [INFO] Usage: scalac <options> <source files>
> [INFO] where possible standard options include:
> [INFO]   -Dproperty=value                     Pass -Dproperty=value directly to the runtime system.
> [INFO]   -J<flag>                             Pass <flag> directly to the runtime system.
> [INFO]   -P:<plugin>:<opt>                    Pass an option to a plugin
> [INFO]   -X                                   Print a synopsis of advanced options.
> [INFO]   -bootclasspath <path>                Override location of bootstrap class files.
> [INFO]   -classpath <path>                    Specify where to find user class files.
> [INFO]   -d <directory|jar>                   destination for generated classfiles.
> [INFO]   -dependencyfile <file>               Set dependency tracking file.
> [INFO]   -deprecation                         Emit warning and location for usages of deprecated APIs.
> [INFO]   -encoding <encoding>                 Specify character encoding used by source files.
> [INFO]   -explaintypes                        Explain type errors in more detail.
> [INFO]   -extdirs <path>                      Override location of installed extensions.
> [INFO]   -feature                             Emit warning and location for usages of features that should be imported explicitly.
> [INFO]   -g:<level>                           Set level of generated debugging info. Choices: (none,source,line,vars,notailcalls), default: vars.
> [INFO]   -help                                Print a synopsis of standard options
> [INFO]   -javabootclasspath <path>            Override java boot classpath.
> [INFO]   -javaextdirs <path>                  Override java extdirs classpath.
> [INFO]   -language:<_,feature,-feature>       Enable or disable language features: `_' for all, `-language:help' to list choices.
> [INFO]   -no-specialization                   Ignore @specialize annotations.
> [INFO]   -nobootcp                            Do not use the boot classpath for the scala jars.
> [INFO]   -nowarn                              Generate no warnings.
> [INFO]   -opt:<_,optimization,-optimization>  Enable optimizations: `_' for all, `-opt:help' to list choices.
> [INFO]   -opt-inline-from:<patterns>          Patterns for classfile names from which to allow inlining, `help` for details.
> [INFO]   -opt-warnings:<_,warning,-warning>   Enable optimizer warnings: `_' for all, `-opt-warnings:help' to list choices.
> [INFO]   -print                               Print program with Scala-specific features removed.
> [INFO]   -release <<release>>                 Compile for a specific version of the Java platform. Supported targets: 6, 7, 8, 9
> [INFO]   -sourcepath <path>                   Specify location(s) of source files.
> [INFO]   -target:<target>                     Target platform for object files. All JVM 1.5 - 1.7 targets are deprecated. Choices: (jvm-1.5,jvm-1.6,jvm-1.7,jvm-1.8), default: jvm-1.8.
> [INFO]   -toolcp <path>                       Add to the runner classpath.
> [INFO]   -unchecked                           Enable additional warnings where generated code depends on assumptions.
> [INFO]   -uniqid                              Uniquely tag all identifiers in debugging output.
> [INFO]   -usejavacp                           Utilize the java.class.path in classpath resolution.
> [INFO]   -usemanifestcp                       Utilize the manifest in classpath resolution.
> [INFO]   -verbose                             Output messages about what the compiler is doing.
> [INFO]   -version                             Print product version and exit.
> [INFO]   @<file>                              A text file containing compiler arguments (options and source files)
> [INFO] 
> [INFO] Deprecated settings:
> [INFO]   -optimise                            Compiler flag for the optimizer in Scala 2.11
> [INFO]                                          deprecated: In 2.12, -optimise enables -opt:l:inline -opt-inline-from:**. Check -opt:help for using the Scala 2.12 optimizer.
> [INFO] 
> [INFO] [INFO] Compiler bridge installed
> ```
> 
> It indicates that the compiler is not invoked correctly here:
> https://github.com/rajcspsg/scala-maven-plugin/blob/d42d0b5c923a55e68ff6e9d33891c1651606cb1c/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L274
> 
> > 1. mvn scala:console is failing with error CNFE: scala.tools.nsc.MainGenericRunner
> 
> In a Scala 2.12 project? If yes, this is probably the same problem than `1.`.

Nope it is for scala 3. I need it for mainConsole configuration https://github.com/rajcspsg/scala_3_mvn_demo/blob/master/pom.xml#L60. For scala2, default value is scala.tools.nsc.MainGenericRunner. I don't know what to configure in case of scala3

rajcspsg commented:
@adpi2 I did remove  ~/.sbt/1.0/zinc/org.scala-sbt/ and tested. But still I get the test-incremental-compile project failing the tests.

adpi2 commented:
> @adpi2 I did remove ~/.sbt/1.0/zinc/org.scala-sbt/ and tested. But still I get the test-incremental-compile project failing the tests.

The `update_for_scala3` branch fails the test but the `master` branch should not. Is it what you get?

So, to fix it you can compare the call [here](https://github.com/rajcspsg/scala-maven-plugin/blob/d42d0b5c923a55e68ff6e9d33891c1651606cb1c/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L274) between both branches and see what's different.

rajcspsg commented:
@adpi2 sorry I didn't clear about the branch specifics in my previous comment.

1. I did `rm -rf  ~/.sbt/1.0/zinc/org.scala-sbt/` and run `mvn clean install` on master. master branch tests are passed.
2. I again did `rm -rf  ~/.sbt/1.0/zinc/org.scala-sbt/` and run `mvn clean install` on  update_for_scala3 branch.  test-incremental-compile integration test failed

Also, did you get chance to check my comment on configuration value for <mainConsole>. Otherwise I get this error https://stackoverflow.com/questions/67075694/whats-the-equivalent-of-scala-tools-nsc-maingenericrunner-in-scala-3?noredirect=1#comment118603047_67075694

adpi2 commented:
> 1. I did rm -rf ~/.sbt/1.0/zinc/org.scala-sbt/ and run mvn clean install on master. master branch tests are passed.
> 2. I again did rm -rf ~/.sbt/1.0/zinc/org.scala-sbt/ and run mvn clean install on update_for_scala3 branch. test-incremental-compile integration test failed

We are on the same page here. The compilation of the Scala 2 bridge is broken in your branch. Maybe you can try to debug it by  printing the values passed [here](https://github.com/rajcspsg/scala-maven-plugin/blob/d42d0b5c923a55e68ff6e9d33891c1651606cb1c/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L274) and [here](https://github.com/rajcspsg/scala-maven-plugin/blob/d42d0b5c923a55e68ff6e9d33891c1651606cb1c/src/main/java/sbt_inc/SbtIncrementalCompiler.java#L312-L319) in both branches. There should be something different that need to be fixed.

> Also, did you get chance to check my comment on configuration value for . Otherwise I get this error https://stackoverflow.com/questions/67075694/whats-the-equivalent-of-scala-tools-nsc-maingenericrunner-in-scala-3?noredirect=1#comment118603047_67075694

Sure, I don't know much this part but I will take a look later this week.

adpi2 commented:
> Nope it is for scala 3. I need it for mainConsole configuration https://github.com/rajcspsg/scala_3_mvn_demo/blob/master/pom.xml#L60. For scala2, default value is scala.tools.nsc.MainGenericRunner. I don't know what to configure in case of scala3
(See my stackoverflow question )

The Scala console is also called the REPL. Its main entrypoint in Scala 3 is [`dotty.tools.repl.Main`](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/repl/Main.scala).

rajcspsg commented:
@adpi2 `dotty.tools.repl.Main` config worked.

rajcspsg commented:
@adpi2 I fixed the incremental compiler issue as well. The root cause is I changed the classifier to null from `"Sources"` based on this [comment](https://github.com/davidB/scala-maven-plugin/pull/447#pullrequestreview-625003316). I reverted now the tests are passing.

rajcspsg commented:
> Do not forgot to complete the contributor sections in root pom.xml with your names (@rajcspsg & @adpi2).
> Thank again for your time and hard work.

@davidB - I've updated @adpi2 and mine in contributors section.

rajcspsg commented:
@davidB - I've fixed all the issues and addressed all the comments. Now we have full pledged maven support for scala3.
Please let me know if there is something that I've missed.

davidB commented on src/main/java/scala_maven/ScalaCompilerSupport.java:
```diff
@@ -171,10 +180,15 @@ protected int compile(
+        compileErrors = true;
+      }
+    } catch (Exception e) {
+      System.out.println("exception compilation error occurred!!!");
+      e.printStackTrace();
```
replace by getLog().warn(...)

davidB replied:
I'll take care of it

davidB commented:
Thanks again for your contribution, I squashed & merged the PR. I'll made few changes (and experiment some possible refactor) before made release.

martin-g commented:
I've just tried 4.4.2-SNAPSHOT (built locally) with my project (that I have recently migrated to SBT 1.5 / Scala 3.0.0-RC2, with `-source:future`. There are two compilation issues with Maven which are not problems with SBT:

1) source code, line 80: `def getHomePage: Class[? <: WebPage] =`
Compilation error: `.../Class1.scala:80: ']' expected but '<:' found.`

2) source code, line 3: `import java.util.{ArrayList as JArrayList, List as JList}`
error: `Class2.scala:3: '}' expected but identifier found.`

martin-g commented:
Update: Today I updated to 3.0.0-RC3 and I realized that scala-maven-plugin was actually still using `<scalaVersion>2.13.5</scalaVersion>`. 🤦🏼 
Everything works fine!

adpi2 commented:
Thanks for testing @martin-g!

davidB commented:
The refactor before release is on-going (you can see it via the draft PR https://github.com/davidB/scala-maven-plugin/pull/459)

I notice a strange things, to be able to build with incremental for scala3, we need to provide scala3-library and scala-library, because both seems to be transitive dependencies of the compiler (and also junit,...)

```sh
./mvnw dependency:tree -f $HOME/.m2/repository/org/scala-lang/scala3-compiler_3.0.0-RC2/3.0.0-RC2/scala3-compiler_3.0.0-RC2-3.0.0-RC2.pom                                                               22:47:37
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building scala3-compiler-bootstrapped 3.0.0-RC2
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ scala3-compiler_3.0.0-RC2 ---
[INFO] org.scala-lang:scala3-compiler_3.0.0-RC2:jar:3.0.0-RC2
[INFO] +- org.scala-lang:scala3-interfaces:jar:3.0.0-RC2:compile
[INFO] +- org.scala-lang:scala3-library_3.0.0-RC2:jar:3.0.0-RC2:compile
[INFO] |  \- org.scala-lang:scala-library:jar:2.13.5:compile
[INFO] +- org.scala-lang:tasty-core_3.0.0-RC2:jar:3.0.0-RC2:compile
[INFO] +- com.novocode:junit-interface:jar:0.11:test
[INFO] |  +- junit:junit:jar:4.11:test
[INFO] |  |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] |  \- org.scala-sbt:test-interface:jar:1.0:test
[INFO] +- org.scala-lang.modules:scala-asm:jar:9.1.0-scala-1:compile
[INFO] +- org.scala-sbt:compiler-interface:jar:1.3.5:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:3.7.0:compile
[INFO] |  \- org.scala-sbt:util-interface:jar:1.3.0:compile
[INFO] +- org.jline:jline-reader:jar:3.19.0:compile
[INFO] +- org.jline:jline-terminal:jar:3.19.0:compile
[INFO] \- org.jline:jline-terminal-jna:jar:3.19.0:compile
[INFO]    \- net.java.dev.jna:jna:jar:5.3.1:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.479 s
[INFO] Finished at: 2021-04-25T22:47:52+02:00
[INFO] Final Memory: 13M/481M
[INFO] ------------------------------------------------------------------------

```

@adpi2 is it expected ?

SethTisue commented:
Yes, that's expected. Scala 3 uses the Scala 2 standard library as-is, and then also adds a small number of additional classes.

adpi2 commented:
It also means that every Scala 3 project depends on the Scala 2.13 standard library.

Even more, they can depend on any other Scala 2.13 libraries. For instance, this will be valid:
```maven
  <properties>
    <scala.version>3.0.0</scala.version>
  </properties>

  <dependencies>
    <dependency>
      <groupId>com.typesafe.akka</groupId>
      <artifactId>akka-http_2.13</artifactId>
      <version>10.2.4</version>
    </dependency>
  </dependencies>
```

davidB commented:
Ok, good to know, I guess the plugin will have some failure with the "check scala version compat" of dependencies.

adpi2 commented:
Regarding Scala version compatibility, the new rules are:
- Starting from 3.0.0, the binary version will be "_3"
- A Scala 3 project can depend on a Scala "_2.13" library
- A Scala 2.13 project can depend on a Scala "_3" library

davidB commented:
I'll do a release without support of new rules about scala version compatibility. 
The mix of scala-library-2.13 and scala3-library will also impact the auto-detection of scala version from dependencies, current workaround, select the highest.
feel free to review PR #459 , I'll merge it before release (planned for next-week)

davidB commented:
FYI, the PR is part of the release 4.5.1
Thanks again

jozic commented:
@davidB looks like 4.5.1 changed behavior in terms of minor Scala versions
see https://github.com/davidB/scala-maven-plugin/discussions/468
